### ISSUE TITLE
Cross-references to topics containing `<shortdesc>` elements with cross-references sometimes result in DOTX031E (file not available) errors

### ISSUE THREAD

**Issue by chrispy-snps** [dita-ot/dita-ot#4221]
## Expected Behavior
If I have a map that references both .dita topics and .md topics:

```
  <topicref keys="topic1" href="topics/topic1.dita"/>
  <topicref keys="topic2" href="topics/topic2.dita"/>
  <topicref keys="warehouse" href="topics/warehouse.dita"/>

  <topicref format="wikidocs" keys="markdown1" href="md/markdown1.md"/>
  <topicref format="wikidocs" keys="markdown2" href="md/markdown2.md"/>
```

and a .dita topic has a keyref cross-reference to an .md topic:

```
<p><xref keyref="markdown1"/></p>
```

the cross-reference should work.

## Actual Behavior
During `topicpull`, errors are issued:

```
[topicpull] Processing file:/tmp/temp20230704180543309/topics/warehouse.dita
[topicpull] file:/.../small/md/markdown1.md:2:24: [DOTX031E][ERROR]: The file markdown2.md is not available to resolve link information.
[topicpull] file:/.../small/md/markdown2.md:2:24: [DOTX031E][ERROR]: The file markdown1.md is not available to resolve link information.
[topicpull] Processing file:/tmp/temp20230704180543309/topics/topic1.dita
[topicpull] file:/.../small/md/markdown1.md:2:24: [DOTX031E][ERROR]: The file markdown2.md is not available to resolve link information.
[topicpull] file:/.../small/md/markdown2.md:2:24: [DOTX031E][ERROR]: The file markdown1.md is not available to resolve link information.
[topicpull] Processing file:/tmp/temp20230704180543309/topics/topic2.dita
[topicpull] file:/.../small/md/markdown1.md:2:24: [DOTX031E][ERROR]: The file markdown2.md is not available to resolve link information.
[topicpull] file:/.../small/md/markdown2.md:2:24: [DOTX031E][ERROR]: The file markdown1.md is not available to resolve link information.
```

The errors occur only when `preprocess` is used. If `preprocess2` is used, the errors do not occur.

The errors occur only when the .dita and .md files are in different subdirectories. If I move them to the root directory, the errors do not occur.

## Possible Solution
Not sure yet.

## Steps to Reproduce
1. Download the following testcase: 
   [4221.zip](https://github.com/dita-ot/dita-ot/files/11952525/4221.zip)
2. Run the following command:
   `dita -i map.ditamap -f html5 -verbose`

## Environment
* DITA-OT version: 4.1
* Operating system and version: Linux (Ubuntu 22.04 LTS)
* How did you run DITA-OT? `dita` command
* Transformation type: `html5`

**Comment by raducoravu** [comment]
At least in my console output there seem to also be some errors like this:

      [topicpull] Failed to transform document: Failed to transform document: Too many nested template or function calls. The stylesheet may be looping.[0m
      [topicpull] Processing file:/../temp/html5/md/markdown1.md

Maybe this is caused by the fact the markdown files refer to each other and adding some logging in the topic pull XSLTs might help...

**Comment by infotexture** [comment]
I'm also seeing these errors when building the DITA-OT docs, but there they seem to stem from Markdown-to-Markdown Xrefs.

From what I've seen so far, the references all seem to work, so the errors may be not be necessary/correct.

**Comment by chrispy-snps** [comment]
@raducoravu - Thanks for mentioning these messages. I noticed them in 4.0.2 but not in 4.1. I am not sure if they are related to this issue or not, but I will look into it.

@infotexture - In fact, that is where I noticed them, while looking into #4218 which used the DITA-OT docs as a testcase.

I will investigate the XSLT processing and follow up here with my findings.

**Comment by chrispy-snps** [comment]
This issue can be reproduced with regular DITA files:

![4221_1](https://github.com/dita-ot/dita-ot/assets/50950969/3b3d5680-4f3d-464d-a0f2-633394ab2767)

When topic1's `<xref>` evaluates topic2's `<shortdesc>`, somehow the nested topic2 `<xref>` evaluation uses topic1's `<topic>` element as the `$baseContextElement`, thus causing the second `<xref>` to attempt to resolve `./topic3.dita` relative to dir1 instead of dir2, thus causing the failure.

Testcase: [4221_tiny.zip](https://github.com/dita-ot/dita-ot/files/11980574/4221_tiny.zip)

The only reason markdown was involved is that markdown topics always use the first paragraphs as a `<shortdesc>` element, thus raising the probability to sensitize this issue.

I will follow up as I know more.

**Comment by raducoravu** [comment]
@chrispy-snps yes, from what I remember as the short description appears as a tooltip when hovering links, it is computed along with the title of the link in the topicpull stage, I see you worked in the past on something similar here: #4184.

**Comment by infotexture** [comment]
@chrispy-snps Thanks for digging deeper on this. The context you provided above might help @jelovirt to figure out the best solution.

**Comment by chrispy-snps** [comment]
I'm still looking into a solution. At first I thought simply using `$linkElement` instead of `$baseContextElement` would handle this case correctly, but that reveals a second issue where sometimes the base URI of `$linkElement` is unexpectedly empty. And the weird part is, the problem occurs intermittently based on the presence or absence of other topics in the job list being processed!

Still trying to figure it out...

**Comment by chrispy-snps** [comment]
@jelovirt - can you mark this issue as fixed when #4239 is merged?

I originally thought "bug 2" above was related to the original issue, but it is a separate issue. I opened #4244 to track it separately.

**Comment by jelovirt** [comment]
@chrispy-snps Sure. For future reference, it's better to edit the issue description to add details about both bug1 and bug2 instead of adding comments about them. That way you don't have to read all the comments in order to understand and reproduce the bug.
